### PR TITLE
Update: relax outerIIFEBody definition (fixes: 6613)

### PR DESCRIPTION
**Pull request by TheCycoONE**
**What issue does this pull request address?**
#6613 

**What changes did you make? (Give an overview)**
Recurse up the AST to the first ExpressionStatement or VariableDeclaration to verify the IIEF is a top level function even if it is not at the top of the AST.

**Is there anything you'd like reviewers to focus on?**
This is my first PR for this project.

### CODE REVIEW

**Comment by TheCycoONE** [comment]
Updated to fix a crash when node.parent.parent was null.

**Comment by gyandeeps** [comment]
Can you add a test for the above case?
Also can you add some invalid test and some test where `outerIIFEBody` is not set to 0?

File: lib/rules/indent.js [hunk 1]
```diff
@@ -367,19 +367,30 @@ module.exports = {
...
         function isOuterIIFE(node) {
             var parent = node.parent;
+            var stmt = parent.parent;
+
+            if (!stmt) {
+                return false;
+            }
+
+            while (
+                stmt.type !== "ExpressionStatement" &&
```
Comment (alberto) [inline lib/rules/indent.js]: Isn't this too relaxing?

**Comment by TheCycoONE** [comment]
@alberto it works for my code base (on files with and without outer IIEFs. I think that every function call must be part of an expression statement or a variable declaration, and that by tracing up the tree we would encounter the statement or declaration before encountering the next block statement. I am new to estree though; so if you have any particular counter examples I'll be willing to add a test for them and resolve the issue.

**Comment by TheCycoONE** [comment]
This patch does generalize an outer IIEF to include any expression that contains an IIEF, e.g. ```!!function(){...}();``` or even ```1 + function(){....}() || 0;``` but I think that such constructs do qualify even if they are not practical. I have heard of the first being used so that truthy returns stay truthy.

**Comment by TheCycoONE** [comment]
Alright, I've thought of a couple counter examples, e.g.:
```js
var o = {
  foo: function() { return true; }()
};
```

I will work on a more robust solution.

**Comment by TheCycoONE** [comment]
I've changed the approach slightly to use a white list of allowed ancestors, using the examples in Ben Alman's blog as a guide: http://benalman.com/news/2010/11/immediately-invoked-function-expression/

The object expression case could be controversial, I have excluded it as an outer IIEF as I think that's the safer choice. Most people are going to be using either the basic ```(function() {...}())``` or a variation of the variations on the module pattern such as ```var ns = function(my) {...}(ns || {});```

**Comment by nzakas** [comment]
Approach looks good. Can you also add some invalid tests to ensure problems will be flagged?

**Comment by TheCycoONE** [comment]
I've added a few invalid cases. I would have added more but I kept running into #1801

**Comment by TheCycoONE** [comment]
Probably would have helped if I had actually committed my changes :)

Invalid tests are now added.

**Comment by ilyavolodin** [comment]
LGTM. Thanks for sticking with this @TheCycoONE

**Pull request merged**